wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/JpaWorkspaceActivityDao.java

```diff
@@ -38,28 +36,170 @@
+
+    Consumer<WorkspaceActivity> update;
+    switch (status) {
+      case RUNNING:
+        update =
+            a -> {
+              a.setStatus(status);
+              a.setLastRunning(timestamp);
+            };
+        break;
+      case STARTING:
+        update =
+            a -> {
+              a.setStatus(status);
+              a.setLastStarting(timestamp);
+            };
+        break;
+      case STOPPED:
+        update =
+            a -> {
+              a.setStatus(status);
+              a.setLastStopped(timestamp);
+            };
+        break;
+      case STOPPING:
+        update =
+            a -> {
+              a.setStatus(status);
+              a.setLastStopping(timestamp);
+            };
+        break;
+      default:
+        throw new ServerException("Unhandled workspace status: " + status);
+    }
+
+    doUpdate(workspaceId, update);
+  }
+
+  @Override
+  @Transactional(rollbackOn = ServerException.class)
+  public List<String> findInStatusSince(long timestamp, WorkspaceStatus status)
+      throws ServerException {
     try {
-      return doFindExpired(timestamp);
+      String queryName = "WorkspaceActivity.get" + firstUpperCase(status.name()) + "Since";
+
+      return managerProvider
+          .get()
+          .createNamedQuery(queryName, WorkspaceActivity.class)
+          .setParameter("time", timestamp)
+          .getResultList()
+          .stream()
+          .map(WorkspaceActivity::getWorkspaceId)
+          .collect(Collectors.toList());
+    } catch (RuntimeException e) {
+      throw new ServerException(e.getLocalizedMessage(), e);
+    }
+  }
+
+  @Override
+  public WorkspaceActivity findActivity(String workspaceId) throws ServerException {
+    try {
+      EntityManager em = managerProvider.get();
+      return em.find(WorkspaceActivity.class, workspaceId);
+    } catch (RuntimeException x) {
+      throw new ServerException(x.getLocalizedMessage(), x);
+    }
+  }
+
+  @Transactional(rollbackOn = ServerException.class)
+  protected void doUpdate(String workspaceId, Consumer<WorkspaceActivity> updater)
+      throws ServerException {
+    doUpdate(false, workspaceId, updater);
+  }
+
+  @Transactional(rollbackOn = ServerException.class)
+  protected void doUpdateOptionally(String workspaceId, Consumer<WorkspaceActivity> updater)
+      throws ServerException {
+    doUpdate(true, workspaceId, updater);
+  }
+
+  private void doUpdate(boolean optional, String workspaceId, Consumer<WorkspaceActivity> updater)
+      throws ServerException {
+    try {
+      EntityManager em = managerProvider.get();
+      WorkspaceActivity activity = em.find(WorkspaceActivity.class, workspaceId);
+      if (activity == null) {
+        if (optional) {
+          return;
+        }
+        activity = new WorkspaceActivity();
+        activity.setWorkspaceId(workspaceId);
+
+        updater.accept(activity);
+
+        em.persist(activity);
```

Comment:
Please add `em.flush` to make sure that persisted data is consistent and there will no any error during transaction commit.

Reply:
Can you explain why that is needed? What is our transaction isolation level?